features/topology-map/org.opennms.features.topology.app/src/test/java/org/opennms/features/topology/app/internal/jung/HierarchyLayoutAlgorithmTest.java

```diff
@@ -36,39 +38,128 @@
+
+    /**
+     * Generates the following graph:
+     *
+     *     t1-v1       t2-v1
+     *    /     \
+     *  t1-v2  t1-v3
+     *   /
+     * t1-v4
+     *
+     * and verifies the resulting X and Y coordinates.
+     */
+    @Test
+    public void notSoBasicLayout() {
+        // Vertices
+        vertices.add(new LevelAwareTestVertex("t1-v1", 0));
+        vertices.add(new LevelAwareTestVertex("t1-v2", 1));
+        vertices.add(new LevelAwareTestVertex("t1-v3", 1));
+        vertices.add(new LevelAwareTestVertex("t1-v4", 2));
+
+        vertices.add(new LevelAwareTestVertex("t2-v1", 0));
+
+        // Edges
+        edges.add(new AbstractEdge("test", "t1-v1-v2", vertices.get(0), vertices.get(1)));
+        edges.add(new AbstractEdge("test","t1-v1-v3", vertices.get(0), vertices.get(2)));
+        edges.add(new AbstractEdge("test","t1-v2-v4", vertices.get(1), vertices.get(3)));
+
+        buildLayout();
+
+        expectVertexToBeAt(0, 150.0d, 100.0d);
+        expectVertexToBeAt(1, 100.0d, 200.0d);
+        expectVertexToBeAt(2, 200.0d, 200.0d);
+        expectVertexToBeAt(3, 100.0d, 300.0d);
+        expectVertexToBeAt(4, 300.0d, 100.0d);
+    }
 
     // NMS-8703
     @Test
     public void verifyHierarchyLayoutFallBack() {
         // Vertices
-        final List<Vertex> vertices = new ArrayList<>();
         vertices.add(new TestVertex("root"));
         vertices.add(new TestVertex("v1"));
         vertices.add(new TestVertex("v2"));
         vertices.add(new TestVertex("v3"));
 
         // Edges
-        final List<Edge> edges = new ArrayList<>();
         edges.add(new AbstractEdge("test", "rootEdge", vertices.get(0), vertices.get(1)));
         edges.add(new AbstractEdge("test","e1", vertices.get(1), vertices.get(2)));
         edges.add(new AbstractEdge("test","e2", vertices.get(2), vertices.get(3)));
 
         // Circle
         edges.add(new AbstractEdge("test","e3", vertices.get(3), vertices.get(1)));
 
+        // Update the layouts and ensure no exception is thrown
+        buildLayout();
+    }
+
+    private void buildLayout() {
         // Mock ALL the things
         final GraphContainer mockGraphContainer = Mockito.mock(GraphContainer.class);
-        final DefaultLayout layout = new DefaultLayout(mockGraphContainer);
+        layout = new DefaultLayout(mockGraphContainer);
         final TestGraph testGraph = new TestGraph(layout, vertices, edges);
         Mockito.when(mockGraphContainer.getGraph()).thenReturn(testGraph);
 
-        // Update layouts and ensure no exception is thrown
+        // Update the layout
         new HierarchyLayoutAlgorithm().updateLayout(mockGraphContainer);
     }
+
+    private void expectVertexToBeAt(int idx, double x, double y) {
+        assertEquals(x, layout.getLocation(vertices.get(idx)).getX(), delta);
+        assertEquals(y, layout.getLocation(vertices.get(idx)).getY(), delta);
+    }
+
+    private static class LevelAwareTestVertex extends TestVertex implements LevelAware {
```

Comment:
Just a thought: Maybe we can create a `GenericLevelAwareVertex` in `topology.api` module.